src/main/java/org/testng/Assert.java

```diff
@@ -118,57 +118,87 @@ public static void assertEquals(Object actual, Object expected, String message)
     assertEqualsImpl(actual, expected, message);
   }
 
+  private static boolean areEqual(Object actual, Object expected){
+    if (expected != null && expected.getClass().isArray()) {
+      return areArraysEqual(actual, expected);
+    }
+    return areEqualImpl(actual, expected);
+  }
+
   /**
    * Differs from {@link #assertEquals(Object, Object, String)} by not taking arrays into special
    * consideration hence comparing them by reference. Intended to be called directly to test
    * equality of collections content.
    */
   private static void assertEqualsImpl(Object actual, Object expected, String message) {
+    boolean equal = areEqualImpl( actual, expected );
+    if(!equal){
+      failNotEquals(actual, expected, message);
+    }
+  }
+
+  private static void assertNotEqualsImpl(Object actual, Object expected, String message) {
+    boolean equal = areEqualImpl( actual, expected );
+    if(equal){
+      failEquals(actual, expected, message);
+    }
+  }
+
+  private static boolean areEqualImpl(Object actual, Object expected){
     if ((expected == null) && (actual == null)) {
-      return;
+      return true;
     }
     if (expected == null ^ actual == null) {
-      failNotEquals(actual, expected, message);
+      return false;
     }
     if (expected.equals(actual) && actual.equals(expected)) {
-      return;
+      return true;
     }
-    fail(format(actual, expected, message, true));
+    return false;
   }
 
-  private static void assertArrayEquals(Object actual, Object expected, String message) {
-    if (expected == actual) {
-      return;
+  /** returns not equal reason or null if equal **/
+  private static String getArrayNotEqualReason(Object actual, Object expected){
+    if(actual == expected) {
+      return null;
     }
-    if (null == expected) {
-      fail("expected a null array, but not null found. " + message);
+    if(null == expected) {
+      return "expected a null array, but not null found";
```

Comment:
@cdalexndr - An invocation to `fail(msg)` would trigger an assertion error with that msg, but now we are returning back error messages, which don't seem to be used anywhere at all. So wouldn't that actually change our error reporting during failures ?

For e.g., `getArrayNotEqualReason()` is being invoked in the below method

```java
private static boolean areArraysEqual(Object actual, Object expected){
     return getArrayNotEqualReason( actual, expected ) == null;
   }
```

but here the error message is not being used at all, which means its getting lost no ?

Reply:
I use the error message to not duplicate code and use a single method for checking equality/inequality. The same method can be used to check equality when error message is null or check inequality when error message is not null.
This method `areArraysEqual` has no usage for the error message because it returns a boolean, but other methods use the error message when failing `fail(errorMsg)`.

Reply:
`getArrayNotEqualReason` error message is used in `getNotEqualDeepReason` method:
```
public static String getNotEqualDeepReason(Set<?> actual, Set<?> expected){
....
  String arrayNotEqualReason = getArrayNotEqualReason( value, expectedValue );
        if(arrayNotEqualReason != null)
          return arrayNotEqualReason;
....
}
```
which is used in `assertEqualsDeep`:
```
  public static void assertEqualsDeep(Set<?> actual, Set<?> expected, String message) {
    String notEqualDeepReason = getNotEqualDeepReason( actual, expected );
    if(notEqualDeepReason != null){
      if (message == null) {
        fail(notEqualDeepReason);
....
```

Reply:
Just changed `getNotEqualDeepReason` method visibility from public to private.